includes/class-moderation.php

```diff
@@ -0,0 +1,334 @@
+		$blocked_keywords = \get_user_meta( $user_id, self::USER_BLOCKED_KEYWORDS_META, true ) ?: array();
+
+		// Extract actor information.
+		$actor_id = '';
+		if ( isset( $activity_data['actor'] ) ) {
+			$actor_id = is_string( $activity_data['actor'] ) ? $activity_data['actor'] : ( $activity_data['actor']['id'] ?? '' );
+		}
+
+		// Check blocked actors.
+		if ( $actor_id && in_array( $actor_id, $blocked_actors, true ) ) {
+			return true;
+		}
+
+		// Check blocked domains.
+		if ( $actor_id ) {
+			$domain = \wp_parse_url( $actor_id, PHP_URL_HOST );
+			if ( $domain && in_array( $domain, $blocked_domains, true ) ) {
+				return true;
+			}
+		}
+
+		// Check blocked keywords in activity content.
+		$activity_json = \wp_json_encode( $activity_data );
+		foreach ( $blocked_keywords as $keyword ) {
+			if ( stripos( $activity_json, $keyword ) !== false ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Add a block for a user.
+	 *
+	 * @param int    $user_id The user ID.
+	 * @param string $type    The block type (actor, domain, keyword).
+	 * @param string $value   The value to block.
+	 * @return bool True on success, false on failure.
+	 */
+	public static function add_user_block( $user_id, $type, $value ) {
+		$meta_key = self::get_user_meta_key_for_type( $type );
+		if ( ! $meta_key ) {
+			return false;
+		}
+
+		$blocks = \get_user_meta( $user_id, $meta_key, true ) ?: array();
+
+		if ( ! in_array( $value, $blocks, true ) ) {
+			$blocks[] = $value;
+			return \update_user_meta( $user_id, $meta_key, $blocks );
+		}
+
+		return true; // Already blocked.
+	}
+
+	/**
+	 * Remove a block for a user.
+	 *
+	 * @param int    $user_id The user ID.
+	 * @param string $type    The block type (actor, domain, keyword).
+	 * @param string $value   The value to unblock.
+	 * @return bool True on success, false on failure.
+	 */
+	public static function remove_user_block( $user_id, $type, $value ) {
+		$meta_key = self::get_user_meta_key_for_type( $type );
+		if ( ! $meta_key ) {
+			return false;
+		}
+
+		$blocks = \get_user_meta( $user_id, $meta_key, true ) ?: array();
+		$key    = array_search( $value, $blocks, true );
+
+		if ( $key !== false ) {
+			unset( $blocks[ $key ] );
+			$blocks = array_values( $blocks ); // Re-index array.
+			return \update_user_meta( $user_id, $meta_key, $blocks );
+		}
+
+		return true; // Not blocked anyway.
+	}
+
+	/**
+	 * Get all blocks for a user.
+	 *
+	 * @param int $user_id The user ID.
+	 * @return array Array of blocks organized by type.
+	 */
+	public static function get_user_blocks( $user_id ) {
+		return array(
+			'actors'   => \get_user_meta( $user_id, self::USER_BLOCKED_ACTORS_META, true ) ?: array(),
+			'domains'  => \get_user_meta( $user_id, self::USER_BLOCKED_DOMAINS_META, true ) ?: array(),
+			'keywords' => \get_user_meta( $user_id, self::USER_BLOCKED_KEYWORDS_META, true ) ?: array(),
```

Comment:
@pfefferle Should actor blocks work in the same way as followers and following, with actor-specific post meta? Or like this, where we keep a list of actor IDs in user meta?

Reply:
good question!?!

It seems that the current version uses whatever you type in the input field, without a lookup. So I would suggest doing that lookup to store the real ID and not any kind of profile URL or WebFinger handle and then we can use the Actor to store the info directly.

Reply:
I think it is also fine to start with the system wide settings and add the user specific ones later on.